qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -266,63 +251,182 @@ def run(self, dag):
         output_layout = Layout({dag.qubits[k]: v for (k, v) in layout_mapping})
         self.property_set["final_layout"] = output_layout
         if not self.fake_run:
-            for node_id in gate_order:
-                node = dag._multi_graph[node_id]
-                process_swaps(
-                    swap_map,
-                    node,
-                    mapped_dag,
-                    original_layout,
-                    canonical_register,
-                    self.fake_run,
-                    self._qubit_indices,
+            mapped_dag = dag.copy_empty_like()
+            _apply_sabre_result(
+                mapped_dag,
+                dag,
+                self._qubit_indices,
+                canonical_register,
+                original_layout,
+                sabre_result,
+            )
+            return mapped_dag
+        return dag
+
+
+def _build_sabre_dag(dag, num_qubits, num_clbits, qubit_indices, clbit_indices):
+    dag_list = []
+    node_blocks = {}
+    for node in dag.topological_op_nodes():
+        cargs = {clbit_indices[x] for x in node.cargs}
+        if node.op.condition is not None:
+            for clbit in dag._bits_in_condition(node.op.condition):
+                cargs.add(clbit_indices[clbit])
+        if isinstance(node.op, ControlFlowOp):
+            node_blocks[node._node_id] = [
+                # TODO: what's the easiest way to get the smallest num_qubits we
+                #  can pass here? A block dag should have fewer qubits than the
+                #  parent, so we could just use that, but really we only need
+                #  max(qubits) + 1
```

Comment:
Still in the early stages of reviewing this. But you can do `block.num_qubits` to get the local number of qubits for a block. But I assume the issue with that is you're relying on the shared indices between the dags (i.e. if inner qubit 0 is outer qubit 5) then you need for `num_qubits` here to be 6. I guess you could do something like `max(qubit_indices[qarg] for qarg in node.qargs) + 1`. I'm not sure how much that'll save us in practice though at the end of the day we only end up saving a few int's in the sabre dag right? It's not like that's critical from a performance PoV.

Reply:
Makes total sense. Yeah that's the issue, the node IDs are in terms of the root DAG, so we need to allocate for all qubits and clbits in the root DAG.

Updated in 29fb9e7.